pubsub-mapped-api/src/main/java/com/google/pubsub/clients/producer/KafkaProducer.java

```diff
@@ -0,0 +1,424 @@
+      } catch (ClassCastException e) {
+        throw new SerializationException("Can't convert value of class " +
+            record.value().getClass().getName() + " to class " +
+            producerConfig.getKafkaConfigs().getClass(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG).getName() +
+            " specified in value.serializer");
+      }
+    }
+
+    if (valueBytes == null || valueBytes.length == 0) {
+      throw new NullPointerException("Value cannot be null or an empty string.");
+    }
+
+    byte[] keyBytes = null;
+    if (record.key() != null) {
+      try {
+        keyBytes = keySerializer.serialize(record.topic(), record.key());
+      } catch (ClassCastException cce) {
+        throw new SerializationException("Can't convert key of class " +
+            record.key().getClass().getName() + " to class " +
+            producerConfig.getKafkaConfigs().getClass(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG).getName() +
+            " specified in key.serializer");
+      }
+    }
+
+    PubsubMessage msg = createMessage(record, dateTime.toString(), keyBytes, valueBytes);
+
+    ApiFuture<String> messageIdFuture = publishers.computeIfAbsent(
+            record.topic(), topic -> createPublisher(topic)).publish(msg);
+
+    final Callback cb = callback;
+    final String topic = record.topic();
+    final int keySize = keyBytes == null ? 0 : keyBytes.length, valueSize = valueBytes.length;
+
+    final SettableFuture<RecordMetadata> future = SettableFuture.create();
+
+    if (callback != null) {
+      if (isAcks) {
+        ApiFutures.addCallback(messageIdFuture, new ApiFutureCallback<String>() {
+          private RecordMetadata recordMetadata =
+              getRecordMetadata(topic, dateTime.getMillis(), keySize, valueSize);
+
+          @Override
+          public void onFailure(Throwable t) {
+            callbackOnCompletion(cb, recordMetadata, new ExecutionException(t));
+            future.setException(t);
+          }
+
+          @Override
+          public void onSuccess(String result) {
+            callbackOnCompletion(cb, recordMetadata, null);
+            future.set(recordMetadata);
+          }
+        });
+      } else {
+        RecordMetadata recordMetadata =
+            getRecordMetadata(topic, -1L, keySize, valueSize);
+        callbackOnCompletion(cb, recordMetadata, null);
+        future.set(recordMetadata);
+      }
+    }
+
+    return future;
+  }
+
+  // Attribute's value's size shouldn't exceed 1024 bytes (256 for key)
+  private PubsubMessage createMessage(ProducerRecord<K, V> record, String dateTime, byte[] key, byte[] value) {
+    Map<String, String> attributes = new HashMap<>();
+
+    attributes.put("id", clientId);
+
+    attributes.put("timestamp", dateTime);
+
+    attributes.put("version", Long.toString(VERSION));
+
+    attributes.put("key", key == null ? "" : new String(Base64.encodeBase64(key)));
+
+    if (attributes.get("key").getBytes().length > 1024) {
```

Comment:
Don't do the getBytes(). This introduces platform-specific encoding. Since we are already base64 encoded, should be able to just get the length of the string.

Reply:
Done.